paho/pinger.go

```diff
@@ -19,120 +19,168 @@ import (
-				if _, err := packets.NewControlPacket(packets.PINGREQ).WriteTo(p.conn); err != nil {
-					if p.pingFailHandler != nil {
-						p.pingFailHandler(err)
-					}
-					return
-				}
-				atomic.AddInt32(&p.pingOutstanding, 1)
-				p.lastPing = time.Now()
-				p.debug.Println("pingHandler sending ping request")
-			}
-		}
+	if p.timer != nil {
+		p.mu.Unlock()
+		return fmt.Errorf("Run() already called")
+	}
+	select {
+	case <-p.done:
+		p.mu.Unlock()
+		return fmt.Errorf("Run() called after stop()")
+	default:
 	}
+	p.keepAlive = keepAlive
+	p.conn = conn
+	p.timer = time.AfterFunc(0, p.sendPingreq) // Immediately send first pingreq
+	p.mu.Unlock()
+
+	return <-p.errChan
+}
+
+func (p *DefaultPinger) Stop() {
+	p.stop(nil)
 }
 
-// Stop is the library provided Pinger's implementation of
-// the required interface function()
-func (p *PingHandler) Stop() {
+func (p *DefaultPinger) PacketSent() {
 	p.mu.Lock()
 	defer p.mu.Unlock()
-	if p.stop == nil {
+	if p.timer == nil {
+		p.debug.Println("PacketSent() called before Run()")
 		return
 	}
-	p.debug.Println("pingHandler stopping")
 	select {
-	case <-p.stop:
-		// Already stopped, do nothing
+	case <-p.done:
+		p.debug.Println("PacketSent() returning due to done channel")
+		return
+	default:
+	}
+
+	p.debug.Println("PacketSent() resetting timer")
+	p.timer.Reset(time.Duration(p.keepAlive) * time.Second)
+}
+
+func (p *DefaultPinger) PingResp() {
+	select {
+	case p.ackReceived <- struct{}{}:
 	default:
-		close(p.stop)
+		p.debug.Println("PingResp() called when ackReceived channel is full")
+		p.stop(fmt.Errorf("received unexpected PINGRESP"))
 	}
 }
 
-// PingResp is the library provided Pinger's implementation of
-// the required interface function()
-func (p *PingHandler) PingResp() {
-	p.debug.Println("pingHandler resetting pingOutstanding")
-	atomic.StoreInt32(&p.pingOutstanding, 0)
+func (p *DefaultPinger) SetDebug(debug log.Logger) {
+	p.debug = debug
+}
+
+func (p *DefaultPinger) sendPingreq() {
+	// Wait for previous ping to be acked before sending another
+	select {
+	case <-p.previousPingAcked:
+	case <-p.done:
+		p.debug.Println("sendPingreq() returning before sending PINGREQ due to done channel")
+		return
+	}
+
+	p.debug.Println("sendPingreq() sending PINGREQ packet")
+	if _, err := packets.NewControlPacket(packets.PINGREQ).WriteTo(p.conn); err != nil {
+		p.stop(fmt.Errorf("failed to send PINGREQ: %w", err))
+		p.debug.Printf("sendPingreq() calling stop() and returning due to packet write error: %v", err)
+		return
+	}
+	p.debug.Println("sendPingreq() sent PINGREQ packet, waiting for PINGRESP")
+	pingrespTimeout := time.NewTimer(time.Duration(p.keepAlive) * time.Second)
+
+	p.PacketSent()
+
+	select {
+	case <-p.done:
+		p.debug.Println("sendPingreq() returning after sending PINGREQ due to done channel")
+	case <-p.ackReceived:
+		p.previousPingAcked <- struct{}{}
+		p.debug.Println("sendPingreq() returning after receiving PINGRESP")
+	case <-pingrespTimeout.C:
```

Comment:
Would generally use `time.After` in this case. `NewTimer` is preferable if you are calling `.Stop` (to release the timer resources; important if you are doing this in a loop). However in this case there will only be one pingrespTimeout in use at a time so I don't think it's really an issue.